p2panda-rs/src/storage_provider/models.rs

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+//! Data models which are used in StorageProvider.
+use std::fmt::Debug;
+
+use crate::document::DocumentId;
+use crate::entry::{decode_entry, EntrySigned, LogId};
+use crate::identity::Author;
+use crate::operation::OperationEncoded;
+use crate::schema::SchemaId;
+use crate::storage_provider::ValidationError;
+use crate::Validate;
+
+/// Struct wrapping an entry with it's operation.
+///
+/// Used internally throughout `storage_provider` in method args and default trait definitions. The
+/// `AsStorageEntry` trait requires `TryFrom<EntryWithOperation>` & `TryInto<EntryWithOperation>`
+/// conversion traits to be present.
+#[derive(Debug, Clone)]
+pub struct EntryWithOperation(EntrySigned, OperationEncoded);
+
+impl EntryWithOperation {
+    /// Instantiate a new EntryWithOperation.
+    pub fn new(entry: &EntrySigned, operation: &OperationEncoded) -> Result<Self, ValidationError> {
+        let entry_with_operation = Self(entry.clone(), operation.clone());
+        entry_with_operation.validate()?;
+        Ok(entry_with_operation)
+    }
+
+    /// Returns a reference to the encoded entry.
+    pub fn entry_encoded(&self) -> &EntrySigned {
+        &self.0
+    }
+
+    /// Returns a refernce to the optional encoded operation.
+    pub fn operation_encoded(&self) -> &OperationEncoded {
+        &self.1
+    }
+}
+
+impl Validate for EntryWithOperation {
+    type Error = ValidationError;
+
+    fn validate(&self) -> Result<(), Self::Error> {
+        self.entry_encoded().validate()?;
+        self.operation_encoded().validate()?;
+        decode_entry(self.entry_encoded(), Some(self.operation_encoded()))?;
+        Ok(())
+    }
+}
+
+/// Struct representing a bamboo append-only log structure.
+#[derive(Debug, Clone)]
+pub struct Log {
```

Comment:
Same as `EntryWithOperation` I could imagine moving `Log` into the core modules as well, as they might be useful for someone else to work with?